### PR TITLE
[hwasan] Add fixed_shadow_base flag

### PR DESCRIPTION
**Pull request by thurstond**
When set to non-zero, the HWASan runtime will map the shadow base at the
specified constant address.

This is particularly useful in conjunction with the existing compiler option
'hwasan-mapping-offset', which bakes a hardcoded constant address into
the instrumentation.

### CODE REVIEW

File: compiler-rt/lib/hwasan/hwasan_flags.inc [hunk 1]
```diff
@@ -84,3 +84,10 @@ HWASAN_FLAG(bool, malloc_bisect_dump, false,
 // are untagged before the call.
 HWASAN_FLAG(bool, fail_without_syscall_abi, true,
             "Exit if fail to request relaxed syscall ABI.")
+
+HWASAN_FLAG(
+    uptr, fixed_shadow_base, 0,
```
Comment (vitalybuka) [inline compiler-rt/lib/hwasan/hwasan_flags.inc:89]: -1, not sure that none will need NULL as a base

**Review by vitalybuka — APPROVED** [review]

File: compiler-rt/lib/hwasan/hwasan_linux.cpp [hunk 2]
```diff
@@ -106,8 +106,14 @@ static uptr GetHighMemEnd() {
 }
 
 static void InitializeShadowBaseAddress(uptr shadow_size_bytes) {
-  __hwasan_shadow_memory_dynamic_address =
-      FindDynamicShadowStart(shadow_size_bytes);
+  // NULL is generally address zero, so it is not a valid location for the
+  // shadow.
+  if (flags()->fixed_shadow_base != 0) {
+    __hwasan_shadow_memory_dynamic_address = flags()->fixed_shadow_base;
```
Comment (vitalybuka) [inline compiler-rt/lib/hwasan/hwasan_linux.cpp:112]: oh,and maybe some test with use of the file?

Comment (vitalybuka) [reply compiler-rt/lib/hwasan/hwasan_linux.cpp:112]: probably linux not android  only test

File: compiler-rt/lib/hwasan/hwasan_flags.inc (hunk 1 above)
Comment (thurstond) [reply compiler-rt/lib/hwasan/hwasan_flags.inc]: Done

File: compiler-rt/lib/hwasan/hwasan_linux.cpp (hunk 2 above)
Comment (thurstond) [reply compiler-rt/lib/hwasan/hwasan_linux.cpp:110]: Added compiler-rt/test/hwasan/TestCases/Linux/fixed-shadow.c

File: compiler-rt/test/hwasan/TestCases/Linux/fixed-shadow.c [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+int main() {
+  __hwasan_enable_allocator_tagging();
+
+  void **mmaps[256];
+  // 48-bit VMA
+  for (int i = 0; i < 256; i++) {
+    unsigned long long addr = (i * (1ULL << 40));
+
+    void *p = mmap((void *)addr, 4096, PROT_READ | PROT_WRITE,
```
Comment (vitalybuka) [inline compiler-rt/test/hwasan/TestCases/Linux/fixed-shadow.c:38]: there is possibility to re-map critical pages with FIXED and crash the process.
Would in't be enough to do the same without FIXED and use addr is a hint?

Alternative trivial approach?
```
print `extern uptr __hwasan_shadow_memory_dynamic_address;`
//CHECK: expected value
```

**Review by vitalybuka — APPROVED** [review]

File: compiler-rt/test/hwasan/TestCases/Linux/fixed-shadow.c (hunk 3 above)
Comment (vitalybuka) [reply compiler-rt/test/hwasan/TestCases/Linux/fixed-shadow.c:42]: > Likewise, this doesn't test the compiler instrumentation.

That's fine to focus on runtime only. E.g. GCC also runs them.

Comment (vitalybuka) [reply compiler-rt/test/hwasan/TestCases/Linux/fixed-shadow.c:42]: > (My concern was that mmap might return addresses that are consecutive pages. In that case, this test will be useless at verifying that the entire address space can be correctly mapped to shadow memory.)

`if ((unsigned long long)p != addr) {` should solve that?

**Review by vitalybuka — APPROVED** [review]

File: compiler-rt/lib/hwasan/hwasan_flags.inc [hunk 4]
```diff
@@ -84,3 +84,10 @@ HWASAN_FLAG(bool, malloc_bisect_dump, false,
 // are untagged before the call.
 HWASAN_FLAG(bool, fail_without_syscall_abi, true,
             "Exit if fail to request relaxed syscall ABI.")
+
+HWASAN_FLAG(
+    uptr, fixed_shadow_base, -1,
```
Comment (vitalybuka) [inline compiler-rt/lib/hwasan/hwasan_flags.inc:89]: @thurstond uptr -> uint64_t 
we can compile for 64bit on 32bit platform

File: compiler-rt/lib/hwasan/hwasan_linux.cpp (hunk 2 above)
Comment (fmayer) [reply compiler-rt/lib/hwasan/hwasan_linux.cpp:110]: For some reason, for static Android binaries, this branch gets taken

**Pull request merged**